websocket.go

```diff
@@ -0,0 +1,277 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/coder/websocket"
+)
+
+type WebSocketProxy struct {
+	upstreamURL string
+	clients     map[*websocket.Conn]map[string]bool // Tracks swap IDs for each client
+	subscribers map[string]map[*websocket.Conn]bool // Tracks clients for each swap ID
+	mu          sync.Mutex
```

Comment:
I wouldn't say we need to do it now, but we should keep in mind the possibility of changing to an RWMutex or even per-swap or per-conn locking in case we see contention issues.